Use generic return type of method to construct proper Jackson writer

danielbobbert opened the pull request:
- Fixes: #43631

gsmet commented on extensions/resteasy-reactive/rest-jackson/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/runtime/serialisers/BasicServerJacksonMessageBodyWriter.java:
```diff
@@ -17,26 +19,46 @@
+            return this.defaultWriter;
+        } else {
+            // compute and cache a specific writer for the given generic type
+            return genericWriters.computeIfAbsent(genericType, type -> {
```
For CHM, it's usually recommended to do a get first when getting a hit is the most common pattern as I was told `computeIfAbsent` locks were a bit less optimized.

danielbobbert replied:
- lambda removed
- get() before computeIfAbsent() added

danielbobbert replied:
I am still not 100% happy with the solution, because I feel I am generating too many custom writers (basically one per different method return type), even though that's probably only necessary for methods that return generic collections etc. (basically all those methods where `method.getGenericReturnType() <> method.getReturnType()). For methods with non-generic return types, the default writer works correctly.
I will dig just a little bit deeper to see whether I can differentiate those cases and only generate a custom writer if it is _really_ necessary.

danielbobbert replied:
I have checked the details of how and why Jackson works this way and performed various other tests. As commented in the code, Jackson needs that custom writer exactly in those cases, where the target type is a parameterized type.
So I have updated the code once more to compute a specialized writer only in those cases, where the method return type actually is a generic (parameterized) type. In all other cases, the default writer can be used.

Imho, the code is good to go.

BTW: I also compared my solution to what is being done in FullyFeaturedServerJacksonMessageBodyWriter (which computes a customer writer per method/type as well). One difference that I spotted is that FullyFeaturedServerJacksonMessageBodyWriter maps the type by name (perTypeWriter in line 40) instead of using the type itself as the map key. I don't see, why this would be beneficial (on the contrary: computing the type name every time should be more expensive that using hashCode() and equals() of the type itself, right?!), but please correct me if I'm wrong.

gsmet reviewed (commented):
Thanks a lot for this PR and all the detective work.

Added a small suggestion and might be a good idea to also avoid the lambda if we can as we try to reduce their usage to the bare minimum in Quarkus runtime code.

danielbobbert committed:
Use generic return type of method to construct proper Jackson writer

gsmet commented:
@danielbobbert just a heads up that we are at Devoxx this week so we will probably have a closer look next week!

geoand reviewed (approved):
Thanks!

geoand commented:
The rest failures seem related

danielbobbert commented:
Yes. The problem is that the "genericType" argument that is passed to the serializer is not properly unwrapped for CompletableFuture. When a method returns a **Uni<X>**, the object "o" passed to writeResponse() is an "X" and the "genericType" is "X.class" (and not Uni<X>). But for a return type of "CompletableFuture<X>", the "o" is of type "X" (so the future has been resolved and the actual value is passed to serialization) but the "genericType" is still "CompletableFuture<X>". Because it is generic, my new code selects a special serializer, but of course that fails, because a serializer for "CompletableFuture<X>" cannot serialize an "X".

danielbobbert commented:
Found it. It's in org.jboss.resteasy.reactive.common.util.types.Types.getEffectiveReturnType()

geoand commented:
Yup, just found that as well. Feel free to push to the change to that method

danielbobbert commented:
Line 185 should probably be changed from 
  `rawType == CompletionStage.class`
to
  `CompletionStage.class.isAssignableFrom(rawType)`
  
Do you want me to add that to my PR and give it a try?

geoand commented:
> Do you want me to add that to my PR and give it a try?

Please do

danielbobbert commented:
OK, two problems:
First off, my isAssignableFrom() check was nonsense since it was doing an additional ".getClass()" on the type itself. I fixed that.
Second, there seems to be another problem, when the response entity is wrapped by a filter, e.g. the HalFilter. I am not 100% sure how to properly fix that. I have added another check that the generic return type is class compliant with the type of the actual object (it would then detect that HalEntity<X>.class != X.class and would use the default writer. What do you think about that?

geoand commented:
> Second, there seems to be another problem, when the response entity is wrapped by a filter, e.g. the HalFilter.

Can you elaborate a little more on this?

Thanks

danielbobbert commented:
Some tests in HalLinksWithJacksonTest were failing. I figured that's because the HalServerResponseFilter wraps the actual entity returned by the method into a HalEntityWrapper (or HalCollectionWrapper).
Now, if my changes in BasicServerJacksonMessageBodyWriter detect that the method was returning a parameterized type (let's call it "X"), they would derive a special writer for "X".
Problem is, that the serializer is then passed a HalEntityWrapper<X> instead of something of type X, and thus fails.
That's why I am now checking that the type of the object being serialized actually matches the genericReturnType of the method, before deriving the specialized writer. If they don't match, I return the default writer as before.

danielbobbert commented:
I wonder, how quarkus-resteasy (legacy) did all of this. Also, I am wondering whether this whole problem also applies to FullyFeaturedServerJacksonMessageBodyWriter. Can you give me a quick hint on how to force Quarkus to use the FullyFeaturedServerJacksonMessageBodyWriter instead of BasicServerJacksonMessageBodyWriter?

geoand commented:
Just use `@JsonView` :)

danielbobbert commented:
Yes, FullyFeaturedServerJacksonMessageBodyWriter also doesn't write type annotations.
Now, should I do the same specialized-writer check in FullyFeaturedServerJacksonMessageBodyWriter as well. Or do you feel am I on the wrong track? I still haven't really gotten around to analyze in depth how the old quarkus-resteasy code solved this. Can you give me some guidance on how to continue?

danielbobbert commented:
OK, I think I found it:
In org.jboss.resteasy.plugins.providers.jackson.ResteasyJackson2Provider (from org.jboss.resteasy:resteasy-jackson2-provider:6.2.9.Final, which is used in quarkus-resteays:3.14.1), lines 162-172, it does more or less the same stuff that I am doing:
```
JavaType rootType = null;
if (genericType != null && value != null && genericType.getClass() != Class.class) {
    rootType = writer.getTypeFactory().constructType(genericType);
    if (rootType.getRawClass() == Object.class) {
        rootType = null;
    }
}

if (rootType != null) {
    writer = writer.forType(rootType);
}
```

One difference seems to be though, that in quarkus-resteasy the "genericType" is only set if the return type of the method is really generic, whereas in quarkus-rest, "genericType" is always set by ServerSerializers.java, lines 198-202:
```
Type genericType;
if (context.hasGenericReturnType()) { // make sure that when a Response with a GenericEntity was returned, we use it
    genericType = context.getGenericReturnType();
} else {
    genericType = target == null ? null : target.getReturnType();
}
```
That's why I had to add the check that "genericType" is indeed a ParameterizedTest.

So I think, I am basically doing the right/intended thing here. I will double check, why/how entity wrapping (e.g. HalResponseFilter) worked in quarkus-resteasy, to finish this up and make sure, that my code resembles the behavior of quarkus-resteasy.

danielbobbert commented:
OK, I think I got it now. Just pushed my latest version which should fix BasicServerJacksonMessageBodyWriter and FullyFeaturedServerJacksonMessageBodyWriter
Can you please give that a try running through CI?

geoand commented:
I launched the run and I'll have a look at the changes tomorrow.

danielbobbert commented:
Ah, I hate it when it does this...
Checks aborted because of unsorted import statements. Sorry for that. I fixed it. Can you please restart the pipeline once more?

geoand commented:
Done 👍

danielbobbert commented:
Thank you!
Oh, and by the way: Could you please point me to the proper documentation how to run the CI myself on my fork? I only found the rather old video linked in Contributing.md. But on my fork, all the steps are always skipped and only the final report is being generated (within seconds). I can't seem to find the proper way to run the tests myself (and also I can't run them locally on my computer because it keeps failing to resolve "org.gradle:gradle-tooling-api:jar:8.1.1" - even though that is definitely installed in my local .m2 repo)

geoand commented:
> But on my fork, all the steps are always skipped and only the final report is being generated (within seconds)

That's odd. IIRC, CI should just run on forks without any necessary intervention on your part.

danielbobbert commented on extensions/resteasy-reactive/rest-jackson/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/runtime/serialisers/ServerJacksonMessageBodyWriter.java:
```diff
@@ -0,0 +1,30 @@
+
+abstract class ServerJacksonMessageBodyWriter extends ServerMessageBodyWriter.AllWriteableMessageBodyWriter {
+
+    ObjectWriter getGenericWriter(Type genericType, ObjectWriter defaultWriter) {
```
OK, I will move the method and also check on those failing tests. Seems like I am still missing a tiny bit...

mariofusco commented on extensions/resteasy-reactive/rest-jackson/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/runtime/serialisers/BasicServerJacksonMessageBodyWriter.java:
```diff
@@ -20,23 +22,42 @@
+    private ObjectWriter getWriter(Type genericType, Object value) {
+        // make sure we properly handle polymorphism in generic collections
+        if (genericType != null && value != null) {
+            ObjectWriter writer = genericWriters.get(genericType);
+            if (writer == null) {
+                // no cached writer for that type. Compute it.
+                writer = JacksonMapperUtil.getGenericWriter(genericType, this.defaultWriter);
+                genericWriters.put(genericType, writer);
+            }
+            return writer;
```
This operation is not atomic and it may happen that multiple writers could be unnecessarily generated for the same type. I believe that this whole block could be replaced with a one-liner, something like:

`return genericWriters.computeIfAbsent(genericType, type -> JacksonMapperUtil.getGenericWriter(type, defaultWriter));`

danielbobbert commented:
OK, so I have double checked with resteasy-classic to solve the problem in the same way as it was done there. Moved the method that derives the appropriate root type to JacksonMapperUtil and call that from both Basic- and FullFeaturedServerJacksonResponseWriter.

This should now pass all tests!

Unfortunately, CI still isn't working for me. But local execution of tests was possible after some fiddling. For some strange reason, tests will not launch correctly every now and then with nonsense ServiceProvider exceptions such as
```
java.lang.RuntimeException: org.junit.jupiter.api.extension.TestInstantiationException: Failed to create test instance
        at io.quarkus.test.QuarkusUnitTest.beforeAll(QuarkusUnitTest.java:736)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
Caused by: org.junit.jupiter.api.extension.TestInstantiationException: Failed to create test instance
        at io.quarkus.test.QuarkusUnitTest.beforeAll(QuarkusUnitTest.java:708)
        ... 1 more
Caused by: java.lang.reflect.InvocationTargetException
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at io.quarkus.test.QuarkusUnitTest.beforeAll(QuarkusUnitTest.java:703)
        ... 1 more
Caused by: java.util.ServiceConfigurationError: io.quarkus.runtime.test.TestHttpEndpointProvider: io.quarkus.resteasy.reactive.server.runtime.ResteasyReactiveTestHttpProvider not a subtype
        at java.base/java.util.ServiceLoader.fail(ServiceLoader.java:593)
        at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1244)
        at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1273)
        at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1309)
        at java.base/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1393)
        at io.quarkus.runtime.test.TestHttpEndpointProvider.load(TestHttpEndpointProvider.java:17)
        at io.quarkus.test.common.http.TestHTTPResourceManager.inject(TestHTTPResourceManager.java:55)
```

geoand commented:
We should also keep in mind that this might have a small performance implication

gsmet commented:
I think it's worth pushing it to 3.16.0, I added the backport label.

geoand commented:
Good idea!

gsmet commented:
@danielbobbert thanks for all your work on this!
